### ISSUE TITLE
Pixel History (and its emulator) does not match what the HW/WARP is rendering on some cases

### ISSUE THREAD

**Issue by darksylinc** [baldurk/renderdoc#3040]
## Description

Debugging through HLSL Pixel Shader, it is evident the GPU (both NV, AMD & WARP) disagree with RenderDoc's emulation of the result.

## Steps to reproduce

1. Open the following capture:  [GpuParticles.rdc.7z.zip](https://github.com/baldurk/renderdoc/files/12444278/GpuParticles.rdc.7z.zip)
2. Go to Event EID 255
3. Debug Pixel [598; 215] 
![RenderDoc](https://github.com/baldurk/renderdoc/assets/3395130/88a0e11b-61b3-42b4-9033-30da86c8c15e)
4. History does not match emulator
   - History says the pixel should be drawn
   - The emulator says the pixel should be discarded: ![RenderDoc2](https://github.com/baldurk/renderdoc/assets/3395130/2a75a398-926b-4531-9e2d-52c771698d2b)
5. Here's another repro using WARP device:  [WARP.rdc.zip](https://github.com/baldurk/renderdoc/files/12444327/WARP.rdc.zip). Try debugging pixel [476; 146] at EID 255

RenderDoc's emulated results is actually what I wanted. Since this works fine on Vulkan but renders improperly on D3D11, I narrowed the culprit down to mod.

~`mod` is [known to behave differently on HLSL vs GLSL](https://forum.unity.com/threads/translating-a-glsl-shader-noise-algorithm-to-hlsl-cg.485750/#post-3164874). Particularly when inputs are negative.~

This means that either:

1. RenderDoc has a divergence (compared to HW) when emulating one of the instructions related to `mod` (like round_ni)
2. RenderDoc has a divergence (compared to HW) when calculating pixScales, which depends on the following functions:
    - exp2
    - log2
    - floor
    - ceil
    - ddx/ddy

~My bug is fixed because I simply have to apply the same mod() formula GLSL uses for HLSL.~
**Update:** My bug was that I implemented a mod() macro in HLSL incorrectly:

```
#define mod( a, b ) (a - b * floor(a / b))

// Should've been:

#define mod( a, b ) ( (a) - (b) * floor( (a) / (b) ) )
```

**This adds to the list of possibly different behavior doing a texture Load out of bounds.**

But RenderDoc is emulating it incorrectly.

## Environment

* RenderDoc version: 1.28
* Operating System: Windows 10
* Graphics API: D3D11
* Tested HW: AMD Radeon RX 560 2GB, NVIDIA GeForce 1060 3GB, WARP

**Comment by Zorro666** [comment]
Hi

The linked file https://github.com/baldurk/renderdoc/files/12444278/GpuParticles.rdc.7z.zip does not appear to match the screenshot you posted which makes the reproduction steps not line up (EID 255 is now VSSetShader).

![image](https://github.com/baldurk/renderdoc/assets/39392/71307559-55cf-4cbd-b608-a95cb06674ff)

Using the WARP reproduction capture.

I am seeing this line in the shader is reading out of bounds:

`ushort materialId	= worldMaterialIdx[inPs.drawId].x & 0x1FFu;`

`inPS.drawId` is 4 and `worldMaterialIdx` only has two elements in it

```
//Uniforms that change per Item/Entity
CONST_BUFFER( InstanceBuffer, 2 )
{
  //.x =
  //The lower 9 bits contain the material's start index.
  //The higher 23 bits contain the world matrix start index.
  //
  //.y =
  //shadowConstantBias. Send the bias directly to avoid an
  //unnecessary indirection during the shadow mapping pass.
  //Must be loaded with uintBitsToFloat
  //
  //.z =
  //lightMask. Ogre must have been compiled with OGRE_NO_FINE_LIGHT_MASK_GRANULARITY
  uint4 worldMaterialIdx[2];
};
```

This could explain a difference in the shader debugger compared to GPU because the shader debugger won't read out of bounds.

In the shader debugger the `materialId` becomes 0 but perhaps it is a different value on the GPU?

For instance element 0 of `worldMaterialIdx` has a materialId of 11 ie.

`worldMaterialIdx[0] 11, 1008981770, 0, 0 0 uint4`

Could this be causing the shader debugger differences you are seeing?

**Comment by darksylinc** [comment]
> The linked file https://github.com/baldurk/renderdoc/files/12444278/GpuParticles.rdc.7z.zip does not appear to match the screenshot you posted which makes the reproduction steps not line up (EID 255 is now VSSetShader).

Ooops. That is a completely unrelated capture. Sorry :face_exhaling: 

> I am seeing this line in the shader is reading out of bounds:
> This could explain a difference in the shader debugger compared to GPU because the shader debugger won't read out of bounds.

Ahhh!!! This is #1084 in action.

TL;DR: We do this on purpose because otherwise shader compilation takes multiple seconds instead of a few ms.

I will turn it off and try again.

**Comment by darksylinc** [comment]
OK, so I just tried again disabling the [2] hack. And:

1. The bug I reported is gone (yay!)
2. A new bug was found (nay...)

Now, RenderDoc history doesn't agree on what is being rendered:

![RenderDoc](https://github.com/baldurk/renderdoc/assets/3395130/59dd7e5e-4c0e-4200-ae44-96181f731e25)

The pixel at EID 255 [677; 178] is clearly rendered as black. The file is actually 7z (Github won't allow uploading .zip files, thus remove the extension)
But RenderDoc history says it was discarded (the debugger at least agrees with the history).
[WARP02.7z.REMOVE_THE_ZIP_EXT.zip](https://github.com/baldurk/renderdoc/files/12448329/WARP02.7z.REMOVE_THE_ZIP_EXT.zip)

The same here (HW capture) EID 255 pixel [1005; 250]. The Pixel is clearly black, but RenderDoc history says it should be grey.  (The file is 7z)

![RenderDoc02](https://github.com/baldurk/renderdoc/assets/3395130/04ae6130-78b1-425f-920f-48e19f865354)

[HW02.rdc.7z.REMOVE_THE_ZIP_EXT.zip](https://github.com/baldurk/renderdoc/files/12448340/HW02.rdc.7z.REMOVE_THE_ZIP_EXT.zip)

I will be changing the title of the ticket to reflect this newly discovered bug.

**Comment by Zorro666** [comment]
After this change the pixel history looks like this (for the WARP case)
![image](https://github.com/baldurk/renderdoc/assets/39392/416b8c6f-102e-4180-9886-d531fe14c395)

**Comment by Zorro666** [comment]
After this change HW02 pixel history looks like this
![image](https://github.com/baldurk/renderdoc/assets/39392/23906415-5de9-4f16-9495-b7baefa8e7d3)
